internal/plugin/connectors/tcp/mssql/connector_test.go

```diff
@@ -49,6 +51,76 @@ func TestHappyPath(t *testing.T) {
 
 	assert.Equal(t, expectedBackendConn, actualBackendConn)
 }
+
+func TestWritePreloginSucceeds(t *testing.T) {
+	logger := logmock.NewLogger()
+	clientConn := mock.NewNetConn(nil)
+	expectedBackendConn := mock.NewNetConn(nil)
+	creds := connector.CredentialValuesByID{
+		"credName": []byte("secret"),
+	}
+
+	// The fields we should send to the user
+	expectedPreloginResponse := map[uint8][]byte{
+		mssql.PreloginVERSION:    {0, 0, 0, 0, 0, 0},
+		mssql.PreloginENCRYPTION: {mssql.EncryptNotSup},
+		mssql.PreloginINSTOPT:    {0},
+		mssql.PreloginTHREADID:   {0, 0, 0, 0},
+		mssql.PreloginMARS:       {0}, // MARS disabled
+	}
+
+	// The fields we should get back from inside the mssql.Connect method
+	mockPreloginResponse := map[uint8][]byte{
+		mssql.PreloginVERSION:    {0, 0, 0, 0, 0, 0},
+		mssql.PreloginENCRYPTION: {mssql.EncryptOn},
+		mssql.PreloginINSTOPT:    {0},
+		mssql.PreloginTHREADID:   {0, 0, 0, 0},
+		mssql.PreloginMARS:       {0}, // MARS disabled
+	}
+
+	// Perform the actual write operation and
+	// assert that we are writing the correct fields back to the client
+	writePreloginWithAssertion := func(
+		tdsBuffer io.ReadWriteCloser,
+		fields map[uint8][]byte,
+		pktType uint8) error {
+			_ = mssql.WritePreloginWithPacketType(tdsBuffer, fields, pktType)
+			assert.Equal(t, expectedPreloginResponse, fields)
+			return nil
+	}
+
+	ctor := mock.NewSuccessfulMSSQLConnectorCtor(
+		func(ctx context.Context) (net.Conn, error) {
+
+			// Pass in our mocked out preloginResponse
+			if preLoginResponse := ctx.Value(ctxtypes.PreLoginResponseKey); preLoginResponse != nil {
+				// A panic will never occur here unless the calling code is wrong
```

Comment:
Doesn't change anything conceptually, but heads up: you'll have to replace `ctxtypes.PreLoginResponseKey` with Kumbi's connectorInterceptor thingy since that will land first thing tomorrow.